### DISCUSSION TITLE
Meta buildpack and detect phases

### DISCUSSION THREAD

**Discussion by till** [paketo-buildpacks/feedback#32] [Q&A]
Hi,

I've googled this for a fair amount of time and read lots of Golang, so I figured I come and ask here if I am on the right track.

The gist is: I am trying to build a meta buildpack with a custom detect which lets me set some variables and what not so users don't have to. Kind of like best practices that shouldn't need configuring by the users.

In order to get there, I started exploring [`pack`](https://github.com/buildpacks/pack), but too much of what I need seems to be in `internal` to make use of it and detect and build are heavily coupled. Then I started exploring `packit` for a bit and managed to build a custom detect phase for applications (kudos for that library).

What I struggle with now is how I would use the result of my own detect phase to assemble already available buildpacks and provide configuration to them. It seems the return of detect is a plan, and that's (I think) one step ahead of what I want to do.

I know that I would need a list of buildpacks I want to use, that's sort of similar to what the `builder` does - not a problem. But the general use of detect seems to be that _my_ buildpack is providing everything needed to build the application, and I would like to avoid that. Instead I would like to for example say:

- this seems to be a nodejs app
- 👉🏼  let's use the node buildpack
- 👉🏼 let's inject the following config values for it

Originally, I wanted a `pack build --dry-run` to run a detect phase and then configure `pack` with `--buildpack` switches and settings. But as I said, I am not getting around in the code base. So I figured, maybe it's easier to use `packit`?

It could be that I am doing this (completely) wrong, so I figured I'd ask.

**Comment by fg-j** [comment]
Hi there! I'll try to shed some light on the situation.

1. It's totally possible (and encouraged!) to combine your buildpack with existing Nodejs buildpacks.

2. Buildpacks collaborate with each other by contributing to the **Build Plan**. Each buildpack contributes to the Build Plan in its detect phase, as you rightly pointed out. As an output to detect, a buildpack will say that it `requires` and/or `provides` certain things. For example. [the `node-engine` buildpack `provides` the node runtime](https://github.com/paketo-buildpacks/node-engine/blob/f8307eb133927b521202d62563b7fb529badc57f/detect.go#L100). Another buildpack can collaborate with node-engine by requiring the node runtime.  The [npm-install buildpack is an example](https://github.com/paketo-buildpacks/npm-install/blob/0f8ba8a8f97e2ee78dab46bfb1e90554c4f353f8/detect.go#L66), since you need node (and npm) to run `npm install`. 

3. Once all buildpacks' detection phases have run, the lifecycle resolves the Build Plan into a **Buildpack Plan** (different, but similarly named!). The Build Plan is a list of requirements and provisions from all of the buildpacks whose detect phases ran. The Buildpack Plan is a deduplicated list of what things are required and which buildpack will be responsible for providing each one. During the build phase, buildpacks look at the Buildpack Plan to see what dependencies they should install, or what commands they should run.

4. So, your custom buildpack might need to use the Build Plan.
    a.  If your custom buildpack relies on other tools to run successfully, it should `require` those.
    b. If the buildpack doesn't need any dependencies to function, it can simply pass detection when the conditions are right, like the [git buildpack](https://github.com/paketo-buildpacks/git/blob/main/detect.go). The buildpack passes detection when a certain service binding is present, and neither provides nor requires anything.

5. To run a build that combines your custom buildpack with other nodejs language family ones, you can either create your own metabuildpack, or add your buildpack alongside the nodejs one at build time.
     a. The Go buildpack is a good [example](https://github.com/paketo-buildpacks/go/blob/main/buildpack.toml) of what it looks like to add a buildpack that sets up config to a composite buildpack. Notice that the git buildpack is an `optional` member of each order group. You could fork the Nodejs buildpack and slot your custom buildpack into its existing order groups in a similar way.
    b. You can also add your custom buildpack to an individual build using the `--buildpack` flag. You'll probably want to include it *before* the Nodejs buildpack, since it sounds like your buildpack needs to configure things before the other Nodejs steps run. Something like:
    ```
    pack build --buildpack my-custom-buildpack --buildpack paketobuildpacks/nodejs:latest
    ```

All in all, I think the [git buildpack](https://github.com/paketo-buildpacks/git) is probably a strong example for you to follow, since it detects on the presence of a certain service binding, provides/requires nothing, and then configures the build environment during its build phase.

**Comment by sophiewigmore** [comment]
> What I struggle with now is how I would use the result of my own detect phase to assemble already available buildpacks and provide configuration to them. It seems the return of detect is a plan, and that's (I think) one step ahead of what I want to do.

For clarity, I'd add on that it sounds like you want an implementation buildpack for the custom detection/configuration settings. The detect phase of your custom implementation buildpack would be responsible for figuring out if you have the right type of app for your use case, requiring anything needed from other buildpacks like @fg-j mentioned, and then in the `build` phase, set whatever configuration is needed. This type of behaviour would be for normal for an implementation buildpack.

Then, the meta-buildpack comes into play when you want to combine your custom buildpack with the pre-existent buildpacks. No custom detection logic would live here, the detection groups are determined by the groups set up in the `buildpack.toml`.

**Reply by sophiewigmore** [reply]
I should add on also that if you're running into problems and want to share - I'd be happy to take a look at what you've put together so far

**Reply by till** [reply]
@sophiewigmore (and @fg-j) thanks so much for your input.

So it seems I may have mis-understood a bit how the phases work (or what they do).

It would seem that my buildpack could be simpler (to stick with an example):

- add my buildpack via `[[ order ]]` groups to a builder
- (in `detect`) probe/test if a `package.json` is in the root of the repository
- then (if `true`), move on to a build phase where I set variables for next buildpacks

The setting variables part is a bit unclear to us right now:
- https://github.com/buildpacks/spec/blob/main/buildpack.md#provided-by-the-buildpacks

Is there an example (with `packit`) where or how that happens?

I looked at the `git` buildpack you mentioned and it seems like my `build` would have to be aware of how the layers I am trying to inject to are called? Or is there a general way to say:

- I detected that this is a node.js app, so I am setting certain variables (for the next builds to follow)
- rely on the `[[ order ]]` groups of the builder that node.js will also be triggered/invoked eventually

Thanks again!

**Reply by sophiewigmore** [reply]
> add my buildpack via [[ order ]] groups to a builder

I think you'd want to add it to the order groups in a language-family buildpack. That would effectively result in your new buildpack detecting alongside the existing buildpacks in the Node language family. Then, if you wanted to create a custom builder with the custom language family buildpack you could, but it wouldn't be necessary. You could still use your custom language family buildpack without the custom builder by passing `--buildpack` in your `pack` command.

> (in detect) probe/test if a package.json is in the root of the repository

Yes, if you want your new buildpack behaviour occur in every instance of a build where a `package.json` is there.

**Reply by sophiewigmore** [reply]
> The setting variables part is a bit unclear to us right now

Yes. This is one of the most confusing things in buildpacks. [This is an example](https://github.com/paketo-buildpacks/node-engine/blob/50f40e17815d9f0cdbc4b580466cc1b8b24ee16b/build.go#L172-L175) of setting an environment variable with `packit`. Whether you set it as a build, launch, or shared environment will determine when the configuration is available.

**Reply by till** [reply]
> > The setting variables part is a bit unclear to us right now
> 
> 
> 
> Yes. This is one of the most confusing things in buildpacks. [This is an example](https://github.com/paketo-buildpacks/node-engine/blob/50f40e17815d9f0cdbc4b580466cc1b8b24ee16b/build.go#L172-L175) of setting an environment variable with `packit`. Whether you set it as a build, launch, or shared environment variable will determine when the configuration is available.

This is perfect. I guess I'll investigate the layers.

Not sure where the best place is, but all the different types of buildpacks - I think it's mostly a matter of linking them up with examples. That would help a lot.

**Reply by fg-j** [reply]
Perhaps more How To docs could help? Instead of just "Create a Paketo Buildpack", more like:
- Create a Basic Buildpack
- Create a Buildpack that Installs Something
- Create a Buildpack that Configures the Build/Launch Environment
etc.

Do you think that'd help you navigate, @till?

**Reply by till** [reply]
Definitely.

I'll publish something too and try to help once I grok it all. 😃

**Reply by sophiewigmore** [reply]
@till I wrote a blog post a while ago that touches on that subject of patterns, but barely scratches the surface. I'll link it here incase it has anything. helpful: https://blog.paketo.io/posts/building-a-cloud-native-buildpack-isnt-as-hard-as-you-think/

**Reply by till** [reply]
@sophiewigmore I (re-)read your blog post on the weekend. I read your blog post a couple weeks (or months) ago before coming on here and after our discussions, this makes a lot more sense! :) Much appreciated.

**Reply by till** [reply]
Re-reading this, I think we misunderstood each other. I know @fg-j mentioned build phase a few times and for lack of understanding, I also overread that or interpreted it wrong.

So what I wanted is, that my "buildpack" sets `BP_`-variables that enable a successful detect of another buildpack. And that is currently not possible.

To give a very concrete example: I can not create a buildpack which parses my nodejs app code and sets `BP_NODE_RUN_SCRIPTS`, so the [node-run-script](https://github.com/paketo-buildpacks/node-run-script) is later detected and used. 

Instead I created a wrapper around [pack](https://buildpacks.io/docs/tools/pack/) to inject `BP_NODE_RUN_SCRIPTS` on that level (before any _detect_ runs).

**Reply by fg-j** [reply]
This use-case makes me wonder if the node-run-script buildpack should always pass detection and simply no-op during the build phase if the `BP_NODE_RUN_SCRIPTS` env var is still not set when its build phase runs. That would allow an earlier buildpack to determine which scripts should run.

**Reply by till** [reply]
Just reading that now. It seems that was my misunderstanding. An earlier buildpack cannot set variables for others. Or would it work if it already passed?

**Reply by fg-j** [reply]
During the detect phase, all buildpacks' `detect` binaries run in parallel with the same inputs. They all have read-only access to the filesystem and the execution evironment. During the build phase, buildpacks' `build` binaries run sequentially. During the build phase, one buildpack can set environment variables for another buildpack's `build` phase. In its `build`, some Buildpack X could look for the presence of some file or attribute of the build and determine that JavaScript assets need to be built. It can set `BP_NODE_RUN_SCRIPTS`. If the  the Node Run Scripts buildpack `build` runs afterwards, it will know which scripts to run.

**Comment by robdimsdale** [comment]
@till That makes sense - sorry it took a couple of attempts to get on the same page.

It sounds like you have a potential future path to solve your specific issue, but I additionally wanted to share that I recently encountered a similar situation as you which wouldn't be solved by supporting node-run scripts.

I was working on a buildpack that would take a function (e.g. a python function) as source code and wrap it inside an application framework (e.g. flask) to create an executable. In principle, this buildpack could be as simple as:

1. wrap the function in a framework
1. write out a package management file (e.g. requirements.txt)
1. write a Procfile with the specified command

And then I could rely on the existing Python buildpack to take this application and process it like any other application. I would add my buildpack to the beginning of the build order, in front of the Python buildpack.

Unfortunately, this does not work as the various Python buildpacks fail detection due to the fact that the source code isn't a valid app (e.g. it is missing a requirements.txt) because the valid application is only present during the Build phase, not during Detect. As you know, the Detect phase runs on the same source code for all buildpacks, and one buildpack cannot influence another buildpack's Detect phase.

I just wanted to share this to validate that while pre-processing / influencing downstream buildpack's detect phase is unusual - and currently unsupported - there are other use-cases for this beyond what you're looking at.

**Reply by till** [reply]
Thanks!

Your use-case sounds interesting as well. E.g. for a knative/faas thing where users provide only a function. It would be great not to have them go learn about bp right away and enable them.

TL;DR - I would really like to have this at some point.

---

The builder RFC mentioned yesterday on the call is going in another interesting direction.

The difference being, if I know what my users are running beforehand, then that makes it easier to configure the builder. But when I don't, it's more guessing.

**Reply by robdimsdale** [reply]
Yup - that's exactly the use case: https://github.com/vmware-tanzu/function-buildpacks-for-knative

**Reply by robdimsdale** [reply]
Also, although pre-processing is somewhat feasible when you're scripting `pack` (or `kpack`), it isn't feasible in the general case when you have a hosted platform like commercial offerings.

**Reply by till** [reply]
So my end goal is arbitrary apps that people wanna run on our platform.
